File: mplad-app/app/pipeline/detectors/d8.py
import sqlite3
import pandas as pd
import json

def run_d8(db_path):
    print("Running D8 (Network)...")
    con = sqlite3.connect(db_path)
    
    works = pd.read_sql_query("SELECT id, district_id, agency_id, title FROM works ORDER BY id", con)
    if works.empty:
        con.close()
        return
        
    detection_rows = []
    
    # Pre-calculate cross_districts per agency
    agency_districts = works.groupby("agency_id")["district_id"].nunique().to_dict()
    
    # Pre-calculate titles per agency per district
    # To quickly find if the same agency used the same title in a different district
    # We can group by (agency_id, title) and collect districts and work_ids
    
    # A dictionary mapping (agency_id, title) -> list of (work_id, district_id)
    ag_title_map = {}
    for _, w in works.iterrows():
        key = (w['agency_id'], w['title'])
        if key not in ag_title_map:
            ag_title_map[key] = []
        ag_title_map[key].append((w['id'], w['district_id']))
        
    for _, w in works.iterrows():
        wid = w['id']
        ag = w['agency_id']
        dist = w['district_id']
        
        cross_dist = agency_districts.get(ag, 1)
        
        # find duplicates
        key = (ag, w['title'])
        matches = ag_title_map.get(key, [])
        
        # filter for OTHER districts
        other_matches = [m_id for m_id, m_dist in matches if m_dist != dist]
        dup_matches = len(other_matches)
        
        # Storage floor: only meaningful rows
        if cross_dist == 1 and dup_matches == 0:
            continue
            
        score = min(1.0, (cross_dist / 5.0) + (dup_matches / 3.0))
        
        evidence = {
            "agency_id": ag,
            "cross_districts": int(cross_dist),
            "dup_matches": int(dup_matches),
            "matched_works": other_matches
        }
        
        detection_rows.append((
            wid,
            "D8",
            float(score),
            json.dumps(evidence)
        ))
        
    con.executemany("INSERT INTO detection_results (work_id, detector, score, evidence_json) VALUES (?, ?, ?, ?)", detection_rows)
    con.commit()
    con.close()

File: mplad-app/lib/pipeline/detectors/d8.js
async function runD8(works, db) {
  console.log('[D8] Running Network Detector...');
  
  const results = [];

  // 1. Calculate cross_districts per agency
  const agencyDistricts = {};
  for (const w of works) {
    if (!w.agency_id || !w.district_id) continue;
    if (!agencyDistricts[w.agency_id]) agencyDistricts[w.agency_id] = new Set();
    agencyDistricts[w.agency_id].add(w.district_id);
  }

  // 2. Map titles to districts for duplicate cross-district titles
  const titleMap = {};
  for (const w of works) {
    if (!w.title || !w.district_id) continue;
    const t = w.title.toLowerCase().trim();
    if (!titleMap[t]) titleMap[t] = [];
    titleMap[t].push(w);
  }

  // 3. Score per work
  for (const w of works) {
    if (!w.agency_id || !w.district_id) continue;
    
    const cross_districts = agencyDistricts[w.agency_id].size;
    
    let dup_matches = 0;
    const matched_works = [];
    
    if (w.title) {
      const t = w.title.toLowerCase().trim();
      const others = titleMap[t] || [];
      for (const o of others) {
        if (o.id !== w.id && o.district !== w.district_id) {
          dup_matches++;
          matched_works.push(o.id);
        }
      }
    }

    const score = Math.min(1.0, (cross_districts / 5.0) + (dup_matches / 3.0));
    
    // Only report if score > 0 (to save DB rows)
    if (score > 0) {
      results.push({
        work_id: w.id,
        detector: 'D8',
        score: score,
        evidence_json: JSON.stringify({
          agency_id: w.agency_id,
          cross_districts: cross_districts,
          dup_matches: dup_matches,
          matched_works: matched_works
        })
      });
    }
  }

  return { results, districtFlags: [] };
}

module.exports = { runD8 };
